### PR TITLE
feat: add IOS searchBar

### PR DESCRIPTION
**Pull request by WoLewicki**
## Description

Added native IOS `SearchBar`.

https://user-images.githubusercontent.com/39658211/117150430-709f2e00-adb8-11eb-8993-9bcc2b6900de.mp4

## Test code and steps to reproduce

See `Test758.tsx` in `TestsExample/` project
or
`SearchBar` playground in `Example/` app

## Checklist

- [x] Included code example that can be used to test this change
- [x] Updated TS types
- [x] Updated documentation: 
  - [x] https://github.com/software-mansion/react-native-screens/blob/master/guides/GUIDE_FOR_LIBRARY_AUTHORS.md
  - [x] https://github.com/software-mansion/react-native-screens/blob/master/native-stack/README.md
  - [x] https://github.com/software-mansion/react-native-screens/blob/master/createNativeStackNavigator/README.md
  - [x] https://github.com/software-mansion/react-native-screens/blob/master/src/types.tsx
  - [x] https://github.com/software-mansion/react-native-screens/blob/master/src/native-stack/types.tsx
- [x] Ensured that CI passes

### CODE REVIEW

**Comment by dioncodes** [comment]
This works perfectly! Can't wait to have this feature in the release version.

**Comment by ha3** [comment]
@WoLewicki great work, thank you. Is this waiting for documentation and type declaration to be updated? If it is so, I'd like to help. Also, I have a question. How can I control the searchbar in the component that is rendered inside the concerning screen? Should I set the searchbar after the component is mounted, like:
```
const Navigator = () => {
  <Stack.Navigator>
    <Stack.Screen name="Component" component={Component} />
  </Stack.Navigator>
}

const Component = () => {
  const ref = useRef();

  useEffect(() => {
    navigation.setOptions({ searchBar: () => <SearchBar ref={ref} /> })
  }, [navigation]);
}
```

Thank you.

**Comment by WoLewicki** [comment]
@ha3 you can see a discussion about it here: https://github.com/software-mansion/react-native-screens/discussions/776 and in #194. I think we need to change the API to not include a component that cannot be used e.g. in Expo managed workflow, so it is not ready yet.

**Comment by Hirbod** [comment]
@WoLewicki I'll check that out, thank you very much.
Have you also considered "SearchView" for android as equivalent?

**Comment by WoLewicki** [comment]
Adding the similar component on Android is a good option for sure and I will probably work on it later. Is there something missing on Android in the component provided by e.g. https://github.com/umhan35/react-native-search-bar? I would probably replicate the already implemented behavior.

**Comment by Hirbod** [comment]
Sorry @WoLewicki I deleted my question and posted it under the discussion (was a mistake posting here)

**Comment by elie-chaaban** [comment]
how to install this specific branch with npm?

**Comment by elie-chaaban** [comment]
This can be used in expo too?

**Comment by dioncodes** [comment]
> how to install this specific branch with npm?

use `"react-native-screens": "software-mansion/react-native-screens#758/head"` in your package.json

**Comment by elie-chaaban** [comment]
i get this

npm ERR! Command failed: git checkout 758/head
npm ERR! error: pathspec '758/head' did not match any file(s) known to git

**Comment by elie-chaaban** [comment]
tried this too
"react-native-screens": "software-mansion/react-native-screens@wolewicki/add-search-bar",


i get  Could not install from "software-mansion\react-native-screens@wolewicki\add-search-bar" as it does not contain a package.json file.

**Comment by dioncodes** [comment]
> i get this
> 
> npm ERR! Command failed: git checkout 758/head
> npm ERR! error: pathspec '758/head' did not match any file(s) known to git

You changed your package json and run `npm install/upgrade`?

**Comment by elie-chaaban** [comment]
of course, i did npm install and now tried upgrade and same error

**Comment by dioncodes** [comment]
> of course, i did npm install and now tried upgrade and same error

It works fine with yarn, maybe try that if this is an option.

**Comment by elie-chaaban** [comment]
i have never used with yarn

**Comment by elie-chaaban** [comment]
anyway, is the search functionnality working on both ios and android?

**Comment by elie-chaaban** [comment]
i have used yarn now but i get 
Invariant Violation: requireNativeComponent: "RNSSearchBar" was not found in the UIManager.

**Comment by dioncodes** [comment]
> i have used yarn now but i get
> Invariant Violation: requireNativeComponent: "RNSSearchBar" was not found in the UIManager.

Run `yarn upgrade`, `cd ios && pod install` and then rebuild your app – that should work

**Comment by Hirbod** [comment]
@elie-chaaban you can't use it on expo, only bare or react native. Once Expo released SDK 41, which will hopefully include this (when it's ready by then), you will be able to use it. 

If you like you could try build it on EXPO using EAS

**Comment by elie-chaaban** [comment]
Pod install? I use windows and expo with react native
Will try yarn update

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by elie-chaaban** [comment]
Oh i use expo the managed workflow, can i use it with expo bare workflow?

**Comment by elie-chaaban** [comment]
In all cases, does this work on android too? 
The whatsapp search toolbar...?

**Comment by elie-chaaban** [comment]
Will try the EAS if it works, i have never tried it yet

**Comment by elie-chaaban** [comment]
Does anyone know how i can manage to get the whatsapp header with the search and everything on ios and android?

**Comment by Hirbod** [comment]
@elie-chaaban this is not a Q&A, it's s pull request. If you got questions use the discussions section or stack overflow

**Comment by ha3** [comment]
@WoLewicki it seems much better now, thank you. Do you have further reservations that block this from merging?

**Comment by WoLewicki** [comment]
I wanted to add the Android implementation before merging this, but it seems like a bit harder task. Maybe you or someone else from this thread would like to make such a PR? I think I will resolve the conflicts and merge this, but having consistent API on both platforms is what we aim for.

**Comment by Hirbod** [comment]
I was hoping this will land into SDK 41... and another 3 months :(

File: native-stack/README.md [hunk 1]
```diff
@@ -256,6 +256,102 @@ Sets the current screen's available orientations and forces rotation if current
...
+    }
+  });
+}, [navigation]);
+```
+
+Supported properties are described below.
+
+#### `autoCapitalize`
+
+Controlles whether text is automatically auto-capitalized as it is entered by the user.
```
Comment (WoLewicki) [inline native-stack/README.md:281]:
```suggestion
Controls whether text is automatically auto-capitalized as it is entered by the user.
```

File: src/index.native.tsx [hunk 2]
```diff
@@ -254,13 +272,19 @@ module.exports = {
...
   },
+  get SearchBar() {
+    if (Platform.OS !== 'ios') {
+      console.warn('Importing SearchBar is only valid on iOS devices.');
+      return View;
+    }
 
-  // these are functions and will not be evaluated until used
-  // so no need to use getters for them
+    return ScreensNativeModules.NativeSearchBar;
```
Comment (WoLewicki) [inline src/index.native.tsx:281]: Why is the previous comment removed?

File: guides/GUIDE_FOR_LIBRARY_AUTHORS.md [hunk 3]
```diff
@@ -209,6 +210,22 @@ Sets the current screen's available orientations and forces rotation if current
 
 Defaults to `default`.
 
+### `searchBar` (iOS only)
```
Comment (WoLewicki) [inline guides/GUIDE_FOR_LIBRARY_AUTHORS.md:213]: There is no such prop in `ScreenStackHeaderConfig`. Please add the correct info about adding a child to `ScreenStackHeaderSearchBarView`, maybe with an example.

**Review by WoLewicki — COMMENTED** [review]
Looks good! I left some comments.

File: guides/GUIDE_FOR_LIBRARY_AUTHORS.md (hunk 3 above)
Comment (kacperkapusciak) [reply guides/GUIDE_FOR_LIBRARY_AUTHORS.md]: Fixed in https://github.com/software-mansion/react-native-screens/pull/758/commits/49247f00af70303e425be0695dba328f3e2b4d4a

File: src/index.native.tsx (hunk 2 above)
Comment (kacperkapusciak) [reply src/index.native.tsx:281]: Fixed in https://github.com/software-mansion/react-native-screens/pull/758/commits/49247f00af70303e425be0695dba328f3e2b4d4a

File: native-stack/README.md (hunk 1 above)
Comment (kacperkapusciak) [reply native-stack/README.md]: Resolved in https://github.com/software-mansion/react-native-screens/pull/758/commits/49247f00af70303e425be0695dba328f3e2b4d4a

**Comment by pouyarezvani** [comment]
@WoLewicki How do I make the SearchBar Always be visible when I enter the page? 
Currently, It's always collapsed. 

https://user-images.githubusercontent.com/32917154/118569767-72a9ab00-b72f-11eb-91d8-c29e16e028cd.mov

**Comment by kacperkapusciak** [comment]
Hi @pouyarezvani 
Sadly, there seem to be a bug in react-native's `ScrollContentView`. You can fix this with this `patch-package` https://github.com/software-mansion/react-native-screens/issues/649#issuecomment-727553714


https://user-images.githubusercontent.com/39658211/118615609-345ccc00-b7c1-11eb-977c-f2aa3444772d.mov

**Comment by pouyarezvani** [comment]
@kacperkapusciak I applied the patch package, however, it only fixed the HeaderLargeTitle collapsed issue on the initial render. now my large headers are not collapsed, but the search bar still is. 
Could you tell me the steps you took to fix the search not being collapsed, please?

**Comment by pouyarezvani** [comment]
@kacperkapusciak I applied the patch package, however, it only fixed the HeaderLargeTitle collapsed issue on the initial render. now my large headers are not collapsed, but the search bar still is. 
Could you tell me the steps you took to fix the search not being collapsed please?

**Comment by pouyarezvani** [comment]
@WoLewicki or @kacperkapusciak 
Screen Shot 2021-05-18 at 4 07 30 PM

Now, I'm getting this error for some reason. 
Going out of a screen with the search bar in it and re-entering it throws me this error. 
Have you guys come across this for any reason? 
Any help would be greatly appreciated!

**Comment by kacperkapusciak** [comment]
@pouyarezvani have you tried setting up `hideWhenScrolling` prop to `false`?

**Comment by pouyarezvani** [comment]
@kacperkapusciak 
Yes, I have but that is not the functionality I'm looking for. I'd like the search bar to hide on scroll. 
Look at the GitHub iOS app for an example. When you enter the home page, the search is always visible, and as you scroll, the search hides.

I just want it to not be collapsed on initially entering that page.

**Comment by pouyarezvani** [comment]
@kacperkapusciak i believe I have,  it I can't remember the result, will check this.

**Comment by kacperkapusciak** [comment]
@pouyarezvani 
you could try setting up `hideWhenScrolling` prop conditionally.
Maybe by setting it by default to `false` and on `setTimeout` or after the user scrolls change it accordingly. 

Good luck with this!

**Comment by pouyarezvani** [comment]
@kacperkapusciak 
Thank you, I know I've been going back and forth on this issue but
I tried adding `onScroll` to the `FlatList`, But it only works if the page re-renders (meaning going to another screen and coming back. 

https://user-images.githubusercontent.com/32917154/119017959-694b5900-b950-11eb-83e5-5e0eb0093eda.mov

I was really hoping there would be a prop on the `searchBar` that would already have this 
behavior: ( initial render not be collapsed, and collapse on scroll - keep in mind I am using the [patch in this comment](https://github.com/software-mansion/react-native-screens/issues/649#issuecomment-727553714) )
I'm just trying to create the behavior that is in iOS GitHub App for their Home Page. 

@WoLewicki Any thoughts on this?

**Comment by kacperkapusciak** [comment]
@pouyarezvani 
I think you've forgotten to include `navigation` in useEffect's dependency array i.e.
```js
useEffect(() => {
 //...
},[navigation, hide])
```
In `react-native-screens` we bridged native search bar functionality and properties to `react-native` for people to use. For additional custom behaviour and props, we are always open to PRs.

Cheers

**Comment by pouyarezvani** [comment]
@WoLewicki Is there any way to change the SearchBar input Text Color? Because I'd like to set it to white on Darkmode. 
Thanks!

**Comment by pouyarezvani** [comment]
@WoLewicki Is there any way to change the SearchBar input Text Color? Because I'd like to set it to any color not dependent on the device theme. 
Thanks!

**Comment by WoLewicki** [comment]
@pouyarezvani please check if  https://github.com/software-mansion/react-native-screens/pull/961 is what you need.

**Comment by pouyarezvani** [comment]
Yes @WoLewicki this is great! 
Thanks!

**Comment by pouyarezvani** [comment]
@WoLewicki is there still no way, without the use of `hideWhenScrolling:false` to render the searchBar not collapsed on the first/initial render of the page? 
I even tried setting `hideWhenScrolling:false` conditionally as someone mentioned, however there was no luck there. 

The patch in the previous comments only fixed the header not this. 

I'd like to not give up hope so quick on this haha.

**Comment by WoLewicki** [comment]
I have no knowledge of how to do it unfortunately, but we are always open to PRs if you find the solution on the internet.

**Comment by Albert-Gao** [comment]
What a great feature, just one quick question, any way I can control the visibility of this search bar, do not want it to be always there.

tried

```typescript
  useEffect(() => {
    if (showSearchBar) {
      navigation.setOptions({
        searchBar: {},
      })
    } else {
      // or plain object
      navigation.setOptions({
        searchBar: undefined,
      })
    }
  }, [navigation, showSearchBar])
```

does not work, is it supported even? or I miss something here? Thanks :)

**Comment by kacperkapusciak** [comment]
Hi @Albert-Gao,
1. If you think that's a bug consider submitting [a new issue](https://github.com/software-mansion/react-native-screens/issues) with a code that reproduces the issue
2. If you've acknowledged the warning you can safely ignore it or disable it by using `LogBox.ignoreLog`
This was described in detail by @WoLewicki in https://github.com/software-mansion/react-native-screens/issues/938

I'm going to lock conversation on this PR because it turned into a Q&A section (and we are using discussions for that 😄 )

**Pull request merged**